File: python/datastructure/otterop/datastructure/linked_list.py
from otterop.lang.panic import Panic as _Panic
from otterop.lang.pure_iterator import PureIterator as _PureIterator
from otterop.lang.string import String as _String
from otterop.lang.oop_object import OOPObject as _OOPObject
from otterop.datastructure.linked_list_node import LinkedListNode as _LinkedListNode
from otterop.datastructure._linked_list_iterator import LinkedListIterator as _LinkedListIterator

class LinkedList:
            
    def __init__(self):
        self._head = None
        self._tail = None
        self._size = 0

    def add_before(self, node, value):
        new_node = _LinkedListNode(value)
        new_node._set_list(self)
        self.add_node_before(node, new_node)
        return new_node

    def _node_of_different_list(self):
        _Panic.invalid_operation(_String.wrap("node of different list"))

    def _remove_on_empty_list(self):
        _Panic.invalid_operation(_String.wrap("remove called on empty list"))

    def add_node_before(self, node, new_node):
        if node.list() != new_node.list() or node.list() != self:
            self._node_of_different_list()
        prev_node = node.prev()
        if prev_node == None:
            new_node.list()._head = new_node
        else:
            prev_node._set_next(new_node)
        new_node._set_prev(prev_node)
        new_node._set_next(node)
        node._set_prev(new_node)
        self._size += 1

    def add_after(self, node, value):
        new_node = _LinkedListNode(value)
        new_node._set_list(self)
        self.add_node_after(node, new_node)
        return new_node

    def add_node_after(self, node, new_node):
        if node.list() != new_node.list() or node.list() != self:
            self._node_of_different_list()
        next_node = node.next()
        if next_node == None:
            new_node.list()._tail = new_node
        else:
            next_node._set_prev(new_node)
        new_node._set_next(next_node)
        new_node._set_prev(node)
        node._set_next(new_node)
        self._size += 1

    def add_first(self, value):
        new_node = _LinkedListNode(value)
        new_node._set_list(self)
        self.add_node_first(new_node)
        return new_node

    def add_node_first(self, new_node):
        if self._head == None:
            if new_node.list() != self:
                self._node_of_different_list()
            self._head = new_node
            self._tail = new_node
            self._size += 1
        else:
            self.add_node_before(self._head, new_node)

    def add_last(self, value):
        new_node = _LinkedListNode(value)
        new_node._set_list(self)
        self.add_node_last(new_node)
        return new_node

    def add_node_last(self, new_node):
        if self._tail == None:
            if new_node.list() != self:
                self._node_of_different_list()
            self._head = new_node
            self._tail = new_node
            self._size += 1
        else:
            self.add_node_after(self._tail, new_node)

    def clear(self):
        self._head = None
        self._tail = None
        self._size = 0

    def remove_first(self):
        if self._head != None:
            self.remove_node(self._head)
        else:
            self._remove_on_empty_list()

    def remove_last(self):
        if self._tail != None:
            self.remove_node(self._tail)
        else:
            self._remove_on_empty_list()

    def remove(self, value):
        curr = self._head
        while curr != None:
            if OOPObject._is(curr.value(), value):
                self.remove_node(curr)
                return True
            curr = curr.next()
        return False

    def remove_node(self, node):
        if node.list() != self:
            _Panic.invalid_operation(_String.wrap("node of different list"))
        prev = node.prev()
        next = node.next()
        if prev != None:
            prev._set_next(next)
        else:
            node.list()._head = next
        if next != None:
            next._set_prev(prev)
        else:
            node.list()._tail = prev
        node._set_prev(None)
        node._set_next(None)
        self._size -= 1

    def size(self):
        return self._size

    def first(self):
        return self._head

    def last(self):
        return self._tail

    def oop_iterator(self):
        return _LinkedListIterator(self)

    def __iter__(self):
        return _PureIterator.new_iterator(self.oop_iterator())

File: dotnet/src/Datastructure/Otterop/Datastructure/LinkedList.cs
namespace Otterop.Datastructure
{
    public class LinkedList<T> : Otterop.Lang.OOPIterable<T>, IEnumerable<T>
    {
        private LinkedListNode<T> head;
        private LinkedListNode<T> tail;
        private int size;
        public LinkedList()
        {
            this.head = null;
            this.tail = null;
            this.size = 0;
        }

        public LinkedListNode<T> AddBefore(LinkedListNode<T> node, T value)
        {
            LinkedListNode<T> newNode = new LinkedListNode<T>(value);
            newNode.SetList(this);
            AddNodeBefore(node, newNode);
            return newNode;
        }

        private void NodeOfDifferentList()
        {
            Otterop.Lang.Panic.InvalidOperation(Otterop.Lang.String.Wrap("node of different list"));
        }

        private void RemoveOnEmptyList()
        {
            Otterop.Lang.Panic.InvalidOperation(Otterop.Lang.String.Wrap("remove called on empty list"));
        }

        public void AddNodeBefore(LinkedListNode<T> node, LinkedListNode<T> newNode)
        {
            if (node.List() != newNode.List() || node.List() != this)
                NodeOfDifferentList();

            LinkedListNode<T> prevNode = node.Prev();
            if (prevNode == null)
                newNode.List().head = newNode;
            else
                prevNode.SetNext(newNode);

            newNode.SetPrev(prevNode);
            newNode.SetNext(node);
            node.SetPrev(newNode);
            this.size++;
        }

        public LinkedListNode<T> AddAfter(LinkedListNode<T> node, T value)
        {
            LinkedListNode<T> newNode = new LinkedListNode<T>(value);
            newNode.SetList(this);
            AddNodeAfter(node, newNode);
            return newNode;
        }

        public void AddNodeAfter(LinkedListNode<T> node, LinkedListNode<T> newNode)
        {
            if (node.List() != newNode.List() || node.List() != this)
                NodeOfDifferentList();

            LinkedListNode<T> nextNode = node.Next();
            if (nextNode == null)
                newNode.List().tail = newNode;
            else
                nextNode.SetPrev(newNode);

            newNode.SetNext(nextNode);
            newNode.SetPrev(node);
            node.SetNext(newNode);
            this.size++;
        }

        public LinkedListNode<T> AddFirst(T value)
        {
            LinkedListNode<T> newNode = new LinkedListNode<T>(value);
            newNode.SetList(this);
            AddNodeFirst(newNode);
            return newNode;
        }

        public void AddNodeFirst(LinkedListNode<T> newNode)
        {
            if (this.head == null)
            {
                if (newNode.List() != this)
                    NodeOfDifferentList();

                this.head = newNode;
                this.tail = newNode;
                this.size++;
            }
            else
            {
                AddNodeBefore(this.head, newNode);
            }

        }

        public LinkedListNode<T> AddLast(T value)
        {
            LinkedListNode<T> newNode = new LinkedListNode<T>(value);
            newNode.SetList(this);
            AddNodeLast(newNode);
            return newNode;
        }

        public void AddNodeLast(LinkedListNode<T> newNode)
        {
            if (this.tail == null)
            {
                if (newNode.List() != this)
                    NodeOfDifferentList();

                this.head = newNode;
                this.tail = newNode;
                this.size++;
            }
            else
            {
                AddNodeAfter(this.tail, newNode);
            }

        }

        public void Clear()
        {
            this.head = null;
            this.tail = null;
            this.size = 0;
        }

        public void RemoveFirst()
        {
            if (this.head != null)
            {
                RemoveNode(this.head);
            }
            else
            {
                RemoveOnEmptyList();
            }

        }

        public void RemoveLast()
        {
            if (this.tail != null)
            {
                RemoveNode(this.tail);
            }
            else
            {
                RemoveOnEmptyList();
            }

        }

        public bool Remove(T value)
        {
            LinkedListNode<T> curr = this.head;
            while (curr != null)
            {
                if (Otterop.Lang.OOPObject.Is(curr.Value(), value))
                {
                    RemoveNode(curr);
                    return true;
                }

                curr = curr.Next();
            }

            return false;
        }

        public void RemoveNode(LinkedListNode<T> node)
        {
            if (node.List() != this)
                Otterop.Lang.Panic.InvalidOperation(Otterop.Lang.String.Wrap("node of different list"));

            LinkedListNode<T> prev = node.Prev();
            LinkedListNode<T> next = node.Next();
            if (prev != null)
            {
                prev.SetNext(next);
            }
            else
            {
                node.List().head = next;
            }

            if (next != null)
            {
                next.SetPrev(prev);
            }
            else
            {
                node.List().tail = prev;
            }

            node.SetPrev(null);
            node.SetNext(null);
            this.size--;
        }

        public int Size()
        {
            return this.size;
        }

        public LinkedListNode<T> First()
        {
            return this.head;
        }

        public LinkedListNode<T> Last()
        {
            return this.tail;
        }

        public Otterop.Lang.OOPIterator<T> OOPIterator()
        {
            return new LinkedListIterator<T>(this);
        }

        public IEnumerator<T> GetEnumerator()
        {
            return Otterop.Lang.PureIterator.NewIterator(OOPIterator());
        }

        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }
    }

}
